### PR TITLE
Refactored to have apply_volume in a specifix mixer

### PR DESCRIPTION
**Pull request by romerod**
This is still work in progress, are you interested in having multiple mixer support?

I started by refactoring the code to have a Mixer trait which is passed to Player.

I verified the code on rpi2 and it works, i'm a complete rust newb so don't hang me for weird code.

TODO:
- call init(), inuse() and release on the mixer
- add possibility for the mixer to notify the player that the volume changed.
- add mixer implementation which allows external process to be responsible for mixing (any ideas for IPC?)

### CODE REVIEW

**Comment by plietar** [comment]
Thanks for working on this, this is an often requested feature.

I don't think the API as it is now would allow for hardware mixers. The `apply_volume` doesn't make any sense in that case.

The best is probably to merge AudioBackend and Mixer. They are closely linked. AudioBackend's which don't support hardware mixing can just fallback to software as it is now.

**Comment by romerod** [comment]
I know that apply_volume doesn't make sense in any but the SoftwareMixer case, that's way the default implementation just returns the unaltered data.

I don't like the idea of putting the audio backend and the mixer together. I've the same use case as @herrernst in #43 where volume will be controlled on an amplifier which I'll remote control with CEC.

Having them separated makes it possible to run different mixers with different backend.

Having the possibility to delegate the "mixing" to an external process would make it possible to use amixer for example without the need to manually implement that.

File: src/mixer/mod.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+use std::borrow::Cow;
+
+pub mod softmixer;
+
+pub trait Mixer {
+    fn init(&mut self);
+    fn inuse(&mut self);
+    fn release(&mut self);
+    fn set(&mut self, volume: u16);
```
Comment (plietar) [inline src/mixer/mod.rs]: Call that `set_volume`

File: src/mixer/mod.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+use std::borrow::Cow;
+
+pub mod softmixer;
+
+pub trait Mixer {
+    fn init(&mut self);
```
Comment (plietar) [inline src/mixer/mod.rs]: Is this necessary ? Initialisation can happen when constructing the `Mixer` object

File: src/mixer/mod.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    fn inuse(&mut self);
```
Comment (plietar) [inline src/mixer/mod.rs]: `start`/`stop` seem like better names

**Comment by plietar** [comment]
Right, I hadn't thought of that use case and it makes more sense now. Thanks for clarifying. I think the API is reasonable like that.

It may make sense the split the `Mixer` trait into a `VolumeControl` and a `VolumeApplier` (terrible names), which share an `Arc<AtomicUsize>`. The former is owned by `SpircManager` and the later by the `Player`. This way the player does not have to worry about volume control, and this can be left to the `SpircManager`.

> - add possibility for the mixer to notify the player that the volume changed.
> - add mixer implementation which allows external process to be responsible for mixing (any ideas for IPC?)

I think this should be left for other PRs. I also have a refactor coming up which would impact the way these are done.

I don't have a candidate for IPC. Is there a precedent for volume control ? Is writing the volume to the process' stdin enough ?

**Comment by romerod** [comment]
I've started splitting it up a bit differently. I introduced the StreamEditor trait which can be optionally passed to the Player. The current volume is owned by the Mixer. Are you ok with this? Currently the state of the device isn't read anymore by the SpircManager [WIP].

File: src/main.rs [hunk 4]
```diff
@@ -120,19 +123,24 @@ fn setup(args: &[String]) -> (Session, Player) {
     matches.opt_str("password"));
     session.login(credentials).unwrap();
 
+     
+    let mixer_name = matches.opt_str("mixer").unwrap_or("SoftMixer".to_owned());
```
Comment (plietar) [inline src/main.rs]: `mixer::find` should take an `Option<&str>`, and take care of choosing a default rather than picking it here, similar to how the audio backend is done.

File: src/main.rs [hunk 5]
```diff
@@ -120,19 +123,24 @@ fn setup(args: &[String]) -> (Session, Player) {
...
+
+    let mixer = mixer::find(&mixer_name).unwrap();
```
Comment (plietar) [inline src/main.rs]: Use `.expect("Invalid mixer")` so it is more obvious in case of an error

File: src/main.rs [hunk 6]
```diff
@@ -120,19 +123,24 @@ fn setup(args: &[String]) -> (Session, Player) {
...
+
     let device_name = matches.opt_str("device");
-    let player = Player::new(session.clone(), move || {
+    let player = Player::new(session.clone(), mixer.get_stream_editor(), move || {
```
Comment (plietar) [inline src/main.rs]: Move `mixer.get_stream_editor()` to a variable

File: src/mixer/mod.rs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+pub mod softmixer;
+
+pub trait Mixer {
+    fn init(&self);
+    fn start(&self);
+    fn stop(&self);
+    fn set_volume(&self, volume: u16);
+    fn volume(&self) -> u16;
+    fn get_stream_editor(&self) -> Option<Box<StreamEditor + Send>>
+    {
```
Comment (plietar) [inline src/mixer/mod.rs]: Brace should go at the end of the line.
If the line is too long, use 
```
fn get_stream_editor(&self)
    -> Option<Box<StreamEditor + Send>>
{
```

File: src/mixer/mod.rs [hunk 8]
```diff
@@ -0,0 +1,26 @@
+use std::borrow::Cow;
+
+pub mod softmixer;
+
+pub trait Mixer {
+    fn init(&self);
```
Comment (plietar) [inline src/mixer/mod.rs]: Why is `init` needed ?

File: src/mixer/mod.rs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    fn set_volume(&self, volume: u16);
+    fn volume(&self) -> u16;
+    fn get_stream_editor(&self) -> Option<Box<StreamEditor + Send>>
+    {
+        None
+    }
+}
+
+pub trait StreamEditor {
+  fn modify_stream<'a>(&self, data: &'a [i16]) -> Cow<'a, [i16]>;
```
Comment (plietar) [inline src/mixer/mod.rs]: Actually I've realised this can take a `&mut [i16]` and modify it in place.
It works out well in the player.

File: src/mixer/mod.rs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+pub fn find(s: &str) -> Option<Box<Mixer + Send>> {
+  match s {
+    "SoftMixer" => Some(Box::new(softmixer::SoftMixer::new())),
```
Comment (plietar) [inline src/mixer/mod.rs]: You can `use softmixer::SoftMixer` at the top level.

File: src/mixer/softmixer.rs [hunk 11]
```diff
@@ -0,0 +1,56 @@
+use super::Mixer;
+use super::StreamEditor;
+use std::borrow::Cow;
+use std::sync::{Arc, RwLock};
+
+pub struct SoftMixer {
+  volume: Arc<RwLock<u16>>
```
Comment (plietar) [inline src/mixer/softmixer.rs]: Make that an atomic.
`AtomicU16` is unstable, but you can use `AtomicUsize` and cast values back and forth

File: src/mixer/softmixer.rs [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        *self.volume.write().unwrap() = volume;
+    }
+    fn get_stream_editor(&self) -> Option<Box<StreamEditor + Send>> {
+        let vol = self.volume.clone();
+        Some(Box::new(SoftVolumeApplier { get_volume: Box::new(move || *vol.read().unwrap() ) }))
+    }
+}
+
+struct SoftVolumeApplier {
+  get_volume: Box<Fn() -> u16 + Send>
```
Comment (plietar) [inline src/mixer/softmixer.rs]: Just store the `Arc<RwLock<u16>>` in there, no need for a closure

File: src/mixer/softmixer.rs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    }
+    fn volume(&self) -> u16 {
+        *self.volume.read().unwrap()
+    }
+    fn set_volume(&self, volume: u16) {
+        *self.volume.write().unwrap() = volume;
+    }
+    fn get_stream_editor(&self) -> Option<Box<StreamEditor + Send>> {
+        let vol = self.volume.clone();
+        Some(Box::new(SoftVolumeApplier { get_volume: Box::new(move || *vol.read().unwrap() ) }))
```
Comment (plietar) [inline src/mixer/softmixer.rs]: That line is too long

File: src/player.rs [hunk 14]
```diff
@@ -47,8 +48,6 @@ pub struct PlayerState {
     pub status: PlayStatus,
     pub position_ms: u32,
     pub position_measured_at: i64,
-    pub update_time: i64,
```
Comment (plietar) [inline src/player.rs:50]: Why did you remove this ?

File: src/player.rs [hunk 15]
```diff
@@ -371,8 +345,11 @@ impl PlayerInternal {
 
                 match packet {
                     Some(Ok(packet)) => {
-                        let buffer = apply_volume(self.state.lock().unwrap().volume,
-                                                  &packet.data);
+                        let buffer = if let Some(ref editor) = stream_editor {
+                            editor.modify_stream(&packet.data)
```
Comment (plietar) [inline src/player.rs]: This could just be `editor.modify_stream(&mut packet.data)` to modify in place.
(See my comment above)

File: src/spirc.rs [hunk 16]
```diff
@@ -43,14 +45,26 @@ struct SpircInternal {
     devices: HashMap<String, String>,
 }
 
+#[derive(Clone)]
+pub struct State {
```
Comment (plietar) [inline src/spirc.rs:48]: Why is this needed ?

**Review by plietar — CHANGES_REQUESTED** [review]
Overall looks good, couple of nits.
`AudioFilter` sounds like a better name than `StreamEditor`. Maybe one day we can have other type of filters.

I'm confused about the changes in `src/spirc.rs` and why they're needed, beyond just adding a `mixer: Box<Mixer + Send>` to the `SpircManager` and calling it when appropriate.

Finally, a device can disable volume changing (probably by not advertising the `kVolumeSteps` in spirc). This could be done if `none` is used as a mixer.
That can be left for another PR though.

File: src/mixer/softmixer.rs (hunk 12 above)
Comment (romerod) [reply src/mixer/softmixer.rs]: There is no reason to allow the SoftVolumeApplier to modify the value, i was looking to be able to get a ReadLock from the RwLock but didn't find one. That's why i decided to use the closure

File: src/player.rs (hunk 14 above)
Comment (romerod) [reply src/player.rs:50]: It's tied to the changes i made in spirc, there is a new struct "State" which contains the whole system state (Player and Mixer) so I thought the update_time belongs to the State not the PlayerState

File: src/spirc.rs (hunk 16 above)
Comment (romerod) [reply src/spirc.rs:48]: see my previous answer about why I removed the update_time, I think when I continue with the changes it will become clear

File: src/main.rs (hunk 4 above)
Comment (romerod) [reply src/main.rs]: done

File: src/main.rs (hunk 5 above)
Comment (romerod) [reply src/main.rs]: done

File: src/main.rs (hunk 6 above)
Comment (romerod) [reply src/main.rs]: done

File: src/mixer/mod.rs (hunk 7 above)
Comment (romerod) [reply src/mixer/mod.rs]: done

File: src/mixer/mod.rs (hunk 10 above)
Comment (romerod) [reply src/mixer/mod.rs]: done

File: src/mixer/softmixer.rs (hunk 13 above)
Comment (romerod) [reply src/mixer/softmixer.rs]: done

File: src/mixer/mod.rs (hunk 9 above)
Comment (romerod) [reply src/mixer/mod.rs]: How can I do that without having issues with the object safety of the AudioFilter trait?
let buffer = if let Some(ref editor) = stream_editor {
^^^^^^ the trait `std::marker::Sized` is not implemented for `[i16]`

File: src/mixer/softmixer.rs (hunk 11 above)
Comment (romerod) [reply src/mixer/softmixer.rs]: Whats the advantage?

Comment (romerod) [reply src/mixer/softmixer.rs]: OK, I just looked it up

Comment (romerod) [reply src/mixer/softmixer.rs]: I get "cannot move out borrowed content" when I use AtomUsize instead of RwLock

File: src/mixer/mod.rs (hunk 9 above)
Comment (plietar) [reply src/mixer/mod.rs]: Does this not work ?

```rust
fn modify_stream(&self, data: &mut [i16]) {
  for x in data {
    *x = (*x as i32 * self.volume as i32 / 0xFFFF) as i16
  }
}

if let Some(ref editor) = stream_editor {
  editor.modify_stream(&mut packet.data)
}
```

(I really don't know why this wasn't done before)

File: src/mixer/softmixer.rs (hunk 11 above)
Comment (plietar) [reply src/mixer/softmixer.rs]: Hmm, you don't need to move out the atomic.
```rust
let volume = self.volume.load(Ordering::Relaxed) as u16;
self.volume.store(volume as usize, Ordering::Relaxed);
```

File: src/mixer/softmixer.rs (hunk 12 above)
Comment (plietar) [reply src/mixer/softmixer.rs]: The privacy boundary is usually the module. As long as it isn't possible to modify the volume unexpectedly from outside the module, that's fine.
I find the closure a lot more confusing than storing the `Arc<RwLock<u16>>`

File: src/mixer/softmixer.rs (hunk 11 above)
Comment (romerod) [reply src/mixer/softmixer.rs]: Done

File: src/mixer/softmixer.rs (hunk 12 above)
Comment (romerod) [reply src/mixer/softmixer.rs]: Done

File: src/mixer/mod.rs (hunk 9 above)
Comment (romerod) [reply src/mixer/mod.rs]: Done, had to copy the packet.data as it's not mutable

File: src/player.rs [hunk 17]
```diff
@@ -371,8 +341,11 @@ impl PlayerInternal {
 
                 match packet {
                     Some(Ok(packet)) => {
-                        let buffer = apply_volume(self.state.lock().unwrap().volume,
-                                                  &packet.data);
+                        let mut buffer = packet.data.to_vec();
```
Comment (plietar) [inline src/player.rs]: Hmm, using `Some(OK(mut packet))` should be enough to avoid the copy.

**Review by plietar — CHANGES_REQUESTED** [review]
Why do you need a channel to send volume change updates back to the `SpircManager` ?
Volume change is only done by spirc anyway, so you can just do `self.notify(false, None);` when a `kMessageTypeVolume` message has been received.

There shouldn't be any changes needed to mercury, and it should definitely not depend on the `spirc` module.
I'm also not sure what the `implement_sender` macro does and why it's needed.
I'm also still not convinced about why the `State` struct in player is needed, and why this existing code wasn't sufficient.

Sorry for requesting all these changes, but I found the implementation very confusing and overly complicated.

File: src/player.rs (hunk 17 above)
Comment (romerod) [reply src/player.rs]: I will try that, thanks

**Comment by romerod** [comment]
There are multiple reasons I used a channel to send volume change notifications:

- The end goal I still have (not for this PR) is to be able to create a mixer which controls an external amplifier, the external amplifier is controllable by other means too, like the remote control

- I didn't want to alter the current way it works. Until now the player was called from the spirc to change the volume, and the player informed the spirc thru the add_observer functionality.

- I didn't want to introduce an additional thread to listen to this changes and I like the way you used the channel to communicate between the spirc and mercury. I extended this approach to be able to share the channel between the spirc and multiple senders, with multiple message types. By following this approach it would be possible to get rid of the Arc<Mutex<>> arround the SpircInternal and just keep a reference to a "CommandSender" and pass a "PlayerStateSender" to the player without week pointers. If you want to take a look at how I came up with the macro take a look at https://repl.it/F1PY/9

I know I modified quite a lot of things, but I didn't want to introduce additional threads and weak pointers.

The dependencies between spirc and mercury were less than ideal, I know, i'd suggest something like the last commit, what do you think? It might be cleaner to move the SpircMessage enum back to spirc, but then the macro would need to create a trait which would need to get boxed to be passed arround etc. I thought it is simpler like that.

Don't worry about asking to much changes, the idea is to get something good out of it, right?

The state part is not finished, TODO on line 435 of spirc.

File: src/player.rs (hunk 17 above)
Comment (romerod) [reply src/player.rs]: done

**Comment by plietar** [comment]
Ok the changes to channel stuff make more sense. It's a shame the `select!` macro isn't stable.
I have a refactor in progress to use tokio and future-rs which would make this a lot simpler though.

I suggest leaving out the volume updates from this PR, especially since it is not needed for software volume. It can be implemented in a much cleaner way once I'm done with the refactor.

**Comment by romerod** [comment]
Hmm, looked at the select! macro, but I don't think it serializes incoming messages as the solution does so you would still have to use locks, which make it much more difficult to know at what point who has to wait on what. Do you see the possible simplification of passing a PlayerStateSender to the instead of the observable stuff and removing the Arc<Mutex<>>?

**Comment by romerod** [comment]
I removed that part of the code (it hurts a bit ;-) ).

As long as the player is playing changes in the mixer will still be reflected as updates are sent to the server, so I can live with that.

**Comment by romerod** [comment]
Is there something missing?

**Comment by sashahilton00** [comment]
@plietar pinging incase of forgotten PR.

File: src/mixer/mod.rs [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    }
+}
+
+pub trait AudioFilter {
+  fn modify_stream(&self, data: &mut [i16]);
+}
+
+pub fn find<T: AsRef<str>>(name: Option<T>) -> Option<Box<Mixer + Send>> {
+  match name {
+    _ => Some(Box::new(SoftMixer::new())),
```
Comment (plietar) [inline src/mixer/mod.rs:21]: Only return SoftMixer if `soft` was given as an argument (or `None` for default).

File: src/player.rs [hunk 19]
```diff
@@ -67,23 +66,20 @@ enum PlayerCommand {
...
     Pause,
-    Volume(u16),
     Stop,
     Seek(u32),
     SeekAt(u32, i64),
 }
 
 impl Player {
-    pub fn new<F>(session: Session, sink_builder: F) -> Player
+    pub fn new<F>(session: Session, stream_editor: Option<Box<AudioFilter + Send>>, sink_builder: F) -> Player
```
Comment (plietar) [inline src/player.rs:75]: Rename the variable to `audio_filter`

File: src/player.rs [hunk 20]
```diff
@@ -229,7 +206,7 @@ fn run_onstop(session: &Session) {
 }
 
 impl PlayerInternal {
-    fn run(self, mut sink: Box<Sink>) {
+    fn run(self, mut sink: Box<Sink>, stream_editor: Option<Box<AudioFilter + Send>>) {
```
Comment (plietar) [inline src/player.rs:209]: Same, change variable name

**Review by plietar — CHANGES_REQUESTED** [review]
Sorry, I thought this was still in WIP.
I still don't understand why you need an extra `State` type ?

**Comment by plietar** [comment]
I've made a few fixes and merged. Thanks @romerod

**Comment by romerod** [comment]
Cool, thank you
